qiskit_nature/components/variational_forms/ucc.py

```diff
@@ -0,0 +1,63 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+TODO.
+"""
+
+from typing import Callable, List, Optional, Tuple, Union
+
+import logging
+
+from .evolved_operator_ansatz import EvolvedOperatorAnsatz
+
+logger = logging.getLogger(__name__)
+
+
+class UCC(EvolvedOperatorAnsatz):
+    """The Unitary Coupled-Cluster Ansatz."""
+
+    def __init__(self, qubit_op_converter: "QubitOpConverter",
+                 num_particles: Optional[Tuple[int, int]] = None,
+                 num_spin_orbitals: Optional[int] = None,
+                 excitations: Optional[Union[str, int, List[int], List[Callable]]] = None):
+        """
+
+        Args:
+            qubit_op_converter: the QubitOpConverter instance which takes care of mapping a
+            :code:`~.SecondQuantizedOp` to a :code:`~.PauliSumOp` as well as performing all
+            configured symmetry reductions on it.
+            num_particles: the tuple of the number of alpha- and beta-spin particles.
+            num_spin_orbitals: the number of spin orbitals.
+            excitations: this can be any of the following:
+                - a `str` which contains the types of excitations. Allowed characters are:
+                    - `s` for singles
+                    - `d` for doubles
+                    - `t` for triples
+                    - `q` for quadruples
+                - a single, positive `int` which denotes the number of excitations (1 == `s`, etc.)
+                - a list of positive integers
+                - a list of `callable` objects which are used to generate the excitations. Each
+                  `callable` must take the following arguments:
+                      - `excitation_type`: the type of excitation to generate
+                      - `num_particles`: the same as above
+                      - `num_spin_orbitals`: the same as above
```

Comment:
About this possibility for specifying the excitations: after discussing with @Cryoris we believe that this is probably the best way to go about introducing flexibility in terms of method variants like what we used to have through the `method_singles` and `method_doubles` parameters. You would then simply have different implementations of the method in this file: https://github.com/Qiskit/qiskit-nature/blob/35fb89c3a579796bf625e8c5dec72a3e1911ee80/qiskit_nature/components/variational_forms/excitation_op_builder.py

While writing this I actually noticed that I did not update that file after the aforementioned discussion. I had originally planned the excitation operator builders as very simple classes rather than standalone functions. I am open for either option but we should discuss what we prefer here. The docstring above kind-of assumes the latter.

@woodsp-ibm @pbark @stefan-woerner please comment :slightly_smiling_face:

Reply:
The QubitConvertor needs either to be the same instance that was used for the problem conversion, so the hopping operators can be symmetry reduced the same way etc, or configured given the same mapper type and the symmetry defined here based on the symmetry sector selected found from the problem. Which means that info needs to be be made available - this used to be put into what was called molecule_info that held this and other values that were needed elsewhere like in UCCSD  ( e.g num particles, orbitals...)

Reply:
I don't see a problem with the `QubitConverter` being the same instance. As per information availability: yes, this is a similar discussion which popped up in #49 at some point. I think it should be sufficient that we have this information available inside of the problem instance. If a "power user" wants to work without the problem instance, it can be reasonably expected that he knows what he is doing and is able to extract the information from the respective places himself.

Reply:
I was not saying anything about whether I considered any way problematic or not, merely that the information needed to get there one way or another and we need to ensure through the design that we have the information, that was all. This is true for UCCSD and also true for HartreeFock initial state that needs consideration/reworking again given the new design.